src/item_collaborative_filtering.py

```diff
@@ -0,0 +1,146 @@
+        """
+
+        filtered_users = df.loc[df[self.item_column] == item, self.user_column]
+        interactions_count = df[df[self.user_column].isin(filtered_users)].groupby(
+            self.user_column,
+            group_keys=False
+        )[self.item_column].agg('nunique').values - 1
+
+        return interactions_count
+
+    def __expected_common_item_pair_users(self, df, item_pair):
+        """
+        :param df: dataframe with columns [user_id, item_id]
+        :param item_pair: tuple(item1, item2)
+        :return: float, expected number of users in common for the item pair (item1, item2)
+        """
+
+        item1, item2 = item_pair[0], item_pair[1]
+
+        product_probability = self.__item_interaction_probability(df, item2)
+
+        interactions_count = self.__count_users_interactions(df, item1)
+
+        return np.sum(1 - (1 - product_probability) ** interactions_count)
+
+    @staticmethod
+    def __recommendations_score_function(expected_users, actual_users):
+        """
+        Given the actual and expected users of two items, it computes how related the items are
+        :param expected_users: np.array
+        :param actual_users: np.array
+        :return:
+        """
+        return (actual_users - expected_users) * np.log(actual_users + 0.1) / np.sqrt(expected_users)
+
+    def fit_recommendations(self, df, item=None, processes=2):
+       """
+       Computes recommendation strength for item pairs
+       By default item=None, means recommendations are computed for all items
+       :param df: dataframe with columns [user_id, item_id]
+       :param item: item that is scored with all other items, if None - all possible item pairs are scored
+       :param processes: (int) number of processes used for parallel computation
+       :return: dataframe with columns [item, recommended_item, actual_common_users, expected_common_users, score]
+       """
+       item_pairs = self.__generate_item_pairs(df, item)
+
+       # output: [((item1, item2), common_users)]
+       # count_pair_users = [self.__count_common_item_pair_users(df, item_pair) for item_pair in item_pairs]
+       count_pair_users = Parallel(n_jobs=processes)(
+           delayed(self.__count_common_item_pair_users)(df, item_pair) for item_pair in item_pairs
+       )
+       # filter out item pairs with no users in common
+       count_pair_users = list(
+           filter(
+           lambda x: x[1] > 0,
+               count_pair_users
+           )
+       )
+
+       # extract item pair, and user count
+       filtered_item_pairs = [item_pair[0] for item_pair in count_pair_users]
+       count_pair_users = [users[1] for users in count_pair_users]
+
+       # output: [expected_users]
+       # compute expected users for item pairs with at least 1 user in common
+       expected_pair_users = Parallel(n_jobs=processes)(
+           delayed(self.__expected_common_item_pair_users)(df, item_pair) for item_pair in filtered_item_pairs
+       )
```

Comment:
Hm... I'm not sure how this parallelization works exactly, but if these runs turn out to be flaky, I'd look if Mutexes need to put in place - for example, what if any of the methods called on `df` actually try to change some internal variables? then this might be problematic if it's happening from a lot of parallel processes..

But if it seems to work fine, then please ignore my comment :)

Reply:
Not sure what you mean 😃 
I need to test the algorithm with some `real` dataset to see if issues pop up.

Reply:
I just wrote a very condensed stream of thoughts here, didn't really explain it well.

I think all is fine if you haven't experienced any parallelization issues so far, but if they ever occur, maybe reconstructing my comment might prove to be marginally useful :P